### ISSUE TITLE
Nanoshooter class work — manage cameras and act as a spawner for spectator and tanks

### ISSUE THREAD

**Issue by chase-moskal** [ChaseMoskal/Nanoshooter#15]
We need an overall solution for managing the active camera when spawning a player-controlled tank, which also implies a new distinction between a player controller tank and an NPC tank.

Living plan:
- `Spectator` entity, can fly around (is the parent class for the upcoming `MapEditor`).
- `Nanoshooter` instance spawns spectator and tanks, sets active cameras, enforces high level game rules, etc.
- Spectator and tank expose a `camera` property for `Nanoshooter` to access.
- `Nanoshooter` should probably have a fallback camera, for when no spectator or anything else is present.
## Criteria
- [ ] `Nanoshooter` instance can spawn player controlled or NPC tanks.
- [ ] `Spectator` entity assumes control whenever no player controlled tanks exist.

**Comment by ChaseMoskal** [comment]
This ticket is done when:
  - Spawner can spawn player controlled or NPC tanks.
  - Spectator entity assumes control whenever no player controlled tanks exist.

This is a precursor to beginning the map editor entity.

**Comment by chase-moskal** [comment]
This ticket is the fulcrum for a lot of functionality, and should be implemented carefully and with wisdom.

In the scope of this ticket, is to formalize the concepts surrounding "Player Controlled" entities — mediating the spawning of player-controlled tanks, and the spawning of a player-controlled spectator cam whenever there aren't any player-controlled tanks.

We need to determine whether having an "active" entity is a kind of mutually-exclusive modal state — can there be only one at a time? Similar to the concept of an "active camera", except an "active entity" for the game world.

That's one approach.

Another approach is to allow entities to decide for themselves when they are "active" or not. Perhaps they publish a Camera, with a precedence value, so that many cameras in the scene may compete for the big rendering opportunity.

We'll probably want to bring back the 'ol Director entity, which is responsible for orchestrating the game's rules, allowing the player to spawn, maintaining the spectator cam and switching to it whenever the player dies.

I think we should code as much of this ticket's functionality into the Director as possible.

I think the Director (naturally) should manage the scene's active camera manually, it spawns tanks, keeps references to them, checks them during its own logic ticks, and is ultimately responsible for removing dead tanks from the game, and during that step, switching back to the spectator cam when the player-controlled tank is removed.

In this case, the director spawns tanks as either `playerControlled: true` or `playerControlled: false`, which determines whether or not the tank accepts inputs, or is an NPC tank which may implement an AI.

**Comment by ChaseMoskal** [comment]
I should formalize a proposal for this, and have it reviewed by a Geoffren.

**Comment by ChaseMoskal** [comment]
Progress continues in a branch named `peterjackson`.

**Comment by chase-moskal** [comment]
I merged and deployed some work that introduced a spectator cam object — which for the first time has us taking control with pointer lock to give us a first person point of view.

**Comment by chase-moskal** [comment]
I've realized that it's too confusing to have an entity which administrates over other entities (Director).

I believe the `Nanoshooter` subclass of `Game` is the simple answer. It should take on all of the responsibilities noted for the supposed "Director".

The `Nanoshooter` class will make good use of toolbox facilities to avoid accumulating bloat.

Most of this ticket's work is done, I just need things officially spawning the way they're supposed to.

I've created these two related tickets:
- #39 — Spectator entity
- #40 — Refactor KeyboardWatcher into InputWatcher
